### PR TITLE
feat(sdk): support dynamic machine type parameters in pipeline task setters

### PR DESCRIPTION
**Pull request by KevinGrantLee**
**Description of your changes:**

Enables dynamic values for cpu_limit, cpu_request, memory_limit, memory_request, accelerator_type, accelerator_limit.

Enables pipelines like
```
@dsl.component
def cpu_limit() -> str:
    return '4000m'

@dsl.component
def sum_numbers(a: int, b:int) -> int:
    return a + b


@dsl.pipeline
def pipeline(
):
    sum_numbers_task = sum_numbers(a=1, b=2)
    sum_numbers_task.set_cpu_limit(cpu_limit().output)
```

**Checklist:**
- [X] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by KevinGrantLee** [comment]
Tests succeed locally, but need to release new kfp-pipeline-spec to merge.

**Comment by gregsheremeta** [comment]
Nice! Is there an open issue for this by any chance? I see it's related to your previous work on this.

**Comment by KevinGrantLee** [comment]
Marking this as draft so we don't accidentally merge it before backend changes are ready. But code is ready to be reviewed. 

Hi Mark, I don't believe we have an open issue for this.

**Comment by KevinGrantLee** [comment]
Backend changes are rolled out. Still need to wait for kfp-pipeline-spec release.

**Comment by KevinGrantLee** [comment]
Hi @convect-bot , yes this sdk change will be able to be run with the current kfp server version. However, you will need to install the next version of kfp-pipeline-spec https://pypi.org/project/kfp-pipeline-spec/ once it is released.

File: sdk/python/test_data/pipelines/pipeline_with_string_machine_fields.py
Comment (chensun) [inline sdk/python/test_data/pipelines/pipeline_with_string_machine_fields.py:1]: This file seems to be largely duplicate with [pipeline_with_resource_spec.py](https://github.com/kubeflow/pipelines/blob/581b7e5b7e888a12652a48f81d9fcd3f5e195e37/sdk/python/test_data/pipelines/pipeline_with_resource_spec.py), let consolidate and merge them if necessary.

Comment (KevinGrantLee) [reply sdk/python/test_data/pipelines/pipeline_with_string_machine_fields.py]: Done, deleted pipeline_with_string_machine_fields

**Review by chensun — APPROVED** [review]
/lgtm

Thank you, @KevinGrantLee !

**Comment by gregsheremeta** [comment]
merging master into your branch resulted in your PR having 33 contributors :joy: 

https://github.com/kubeflow/pipelines/commit/94b1a0de73e7d291d5f0581e92b8d637a335964f

weird eh?

@chensun have you seen that before? How do we avoid that kind of thing?

**Comment by HumairAK** [comment]
Yeah I think it's good form to just have the author rebase and keep the commits clean in the PR (reviewers and approvers should keep the author honest here) 
The approver could squash/rebase/clean up, but then they will end up being co-author. Better to just have the author do it if feasible and non urgent.

**Comment by gregsheremeta** [comment]
+1 to asking author to rebase

**Comment by ougni** [comment]
Hi Team, 

I've tried to compile the following pipeline but I'm still getting the error below. I'd really appreciate if someone can point me to the correct usage here. 

```python 
import kfp
from kfp import dsl


@dsl.component
def cpu_limit() -> str:
    return "4000m"


@dsl.component
def sum_numbers(a: int, b: int) -> int:
    return a + b


@dsl.pipeline
def pipeline():
    sum_numbers_task = sum_numbers(a=1, b=2)
    sum_numbers_task.set_cpu_limit(cpu_limit().output)


kfp.compiler.Compiler().compile(
    pipeline_func=pipeline,
    package_path="test.json",
)
```
```
.env/lib/python3.11/site-packages/kfp/compiler/compiler_utils.py", line 827, in validate_cpu_request_limit_to_float
    raise ValueError(
ValueError: Invalid cpu string. Should be float or integer, or integer followed by "m".
```


KFP Versions:
```
kfp                              2.11.0
kfp-pipeline-spec                0.6.0
kfp-server-api                   2.3.0
```

**Comment by zackgow** [comment]
@ougni downgrading kfp to 2.10.0 worked for me, so this appears to be a compiler bug in the 2.11.0 release.

> 
> KFP Versions:
> 
> ```
> kfp                              2.11.0
> kfp-pipeline-spec                0.6.0
> kfp-server-api                   2.3.0
> ```

**Pull request merged**
